Session: Implement session build transactions.

ikmsk10 opened the pull request:
In some cases creating a session might be from different services. It
requires to be a transaction pattern.
This patch brings the implementation of creating a session in
a transaction and a little bit of refactoring.

End-user-impact: None

Signed-off-by: Igor Kononenko <i.kononenko@yadro.com>

ikmsk10 committed:
Session: Implement session build transactions.

In some cases creating a session might be from different services. It
requires to be a transaction pattern.
This patch brings the implementation of creating a session in
a transaction and a little bit of refactoring.

End-user-impact: None

Signed-off-by: Igor Kononenko <i.kononenko@yadro.com>

AlexanderAmelkin reviewed (changes requested):
Waiting for API changes as agreed

ikmsk10 committed:
api: Move to the C-style api

Many services are implemented on the C-lang. So we need to support both
C/C++ public API to access session manager features.

End-user-impact: None

Signed-off-by: Igor Kononenko <i.kononenko@yadro.com>

nest1ing commented on include/libobmcsession/obmcsession.h:
```diff
@@ -0,0 +1,181 @@
+
```
This empty line is excess.

nest1ing commented on include/libobmcsession/obmcsession.hpp:
```diff
@@ -0,0 +1,77 @@
+
```
It seems to me this header is excess at all.
The C++ app could use the C-style API.

ikmsk10 replied:
In some cases C++ services use sdbusplus::bus::bus and it is too difficult to retrieve the sd_bus-C-style handle to pass into the C-style API functions of the current library. So, I prefer to define both API where it required,

nest1ing replied:
Yep, `sdbusplus::bus` has `get()` method, but unfortunately, it's protected.

nest1ing commented on src/api.cpp:
```diff
@@ -0,0 +1,301 @@
+static SessionManagerPtr managerPtr;
+static std::unique_ptr<sdbusplus::bus::bus> dbus;
+
+int obmcsesManagerInit(sd_bus* bus, const char* slug, uint8_t type);
```
All these forward declarations are also excess. Just include required header.

ikmsk10 replied:
I've looked at the modern libraries, and this is a general style that I've picked from it. But there isn't matter , I might remove the declaration in.cpp

nest1ing replied:
As I can see, these functions don't use each other and the forward declarations or including the header seem unnecessary at all.

ikmsk10 replied:
Ok, I'll fix that.

nest1ing commented on src/api.cpp:
```diff
@@ -0,0 +1,301 @@
+    }
+    try
+    {
+        dbus = std::make_unique<sdbusplus::bus::bus>(bus);
```
As the separate C++ API is excess, I suggest to pass the raw `sd_bus` pointer to the SessionManager constructor.
In addition, `sdbusplus::bus` is kind of smart pointer and it doesn't require the extra wrapping into `std::uniq_ptr`.

ikmsk10 replied:
I've still beliave the Cpp impelentation of serveral API-functions is required to cleanly provide sdbusplus handles.

nest1ing replied:
Ok, the C++ API is also required. But keeping the bus object wrapped into `std::uniq_ptr` seems to me like overhead.
This still could be a member of `SessionManager` that might have two constructors.
The first gets the raw pointer and create `sdbusplus::bus` object from it,
the second just copies the bus object from C++ app.

ikmsk10 replied:
Yes, it makes sense. I'll add another constructor with sd_bus. Thanks for the proposal.

nest1ing commented on include/libobmcsession/obmcsession.hpp:
```diff
@@ -0,0 +1,77 @@
+ * @return int                      - session build status
+ */
+extern int obmcsesCreateWithFCleanup(const std::string& userName,
+                                     uint32_t remoteAddress,
```
Seems I missed it in the previous review.
What about IPv6 addresses?

ikmsk10 replied:
Hmm, is there should be boost:asio::ip:address? I've very dislaked include boost into the current library only to define ip-address. Perhaps i can define new struct to store IPv4/IPv6 (value,type)?

nest1ing replied:
Are you sure that it should be stored in binary format?
Can we store it as a string?

ikmsk10 replied:
Perhaps You are right. I've preferred to use trivial types where possible, and the numeric comparison is faster, but it would be easier to store address as a string now.

AlexanderAmelkin commented on include/libobmcsession/obmcsession.h:
```diff
@@ -0,0 +1,181 @@
+ */
+
+#ifndef __OBMCSESSION_H__
+#define __OBMCSESSION_H__
```
better use `#pragma once`.

AlexanderAmelkin commented on include/libobmcsession/obmcsession.h:
```diff
@@ -0,0 +1,181 @@
+ *                            items from. Service name template:
+ *                            'xyz.openbmc_project.Session.${slug}'.
+ * @param[in] type          - Type of all session items that will be created
+ * by the current instance.
```
please align

AlexanderAmelkin commented on include/libobmcsession/obmcsession.h:
```diff
@@ -0,0 +1,181 @@
+ *                            'xyz.openbmc_project.Session.${slug}'.
+ * @param[in] type          - Type of all session items that will be created
+ * by the current instance.
+ * @return int              - session manager initialization status
```
Please specify that it's an `errno-like status`, not just `status`. Now it looks like it returns a zero/non-zero (which is also true), but nobody could ever guess that it actually returns an errno value.
Alternatively you could specify:
```
@retval == 0 - Success
@retval != 0 - An errno-like value indicating the type of failure
```

AlexanderAmelkin commented on include/libobmcsession/obmcsession.h:
```diff
@@ -0,0 +1,181 @@
+ *
+ * @note The session is incomplete. Metadata of session must be set for 10
+ *       seconds. Otherwise, the current session will be closed. All other
+ *       new session creates requests will be rejected until the current
```
`creates` -> `create` ?

AlexanderAmelkin commented on include/libobmcsession/obmcsession.h:
```diff
@@ -0,0 +1,181 @@
+ *
+ * @param[out] sessionId    - unique session ID
+ *
+ * @note The session is incomplete. Metadata of session must be set for 10
```
"The _created_ session is incomplete" ?
"must be set _within_ 10 seconds" ?

AlexanderAmelkin commented on include/libobmcsession/obmcsession.h:
```diff
@@ -0,0 +1,181 @@
+
+/**
+ * @brief Create a session and publish into the dbus with cleanup callback
+ *        on the session destroy.
```
"upon session destruction".

AlexanderAmelkin commented on include/libobmcsession/obmcsession.h:
```diff
@@ -0,0 +1,181 @@
+ * @param[in] remoteAddress         - the IP address of the session
+ *                                    initiator.
+ * @param[in] cleanupFn             - the callback to cleanup session on
+ *                                    destroy.
```
"the session",
"upon destruction"

AlexanderAmelkin commented on include/libobmcsession/obmcsession.h:
```diff
@@ -0,0 +1,181 @@
+                               obmcsesSessionId* sessionId);
+
+/** @brief Commit the pending session build that is handle by the current
+ *         manager.
```
"session creation" ?
"handled"

AlexanderAmelkin commented on include/libobmcsession/obmcsession.h:
```diff
@@ -0,0 +1,181 @@
+/** @brief Commit the pending session build that is handle by the current
+ *         manager.
+ *
+ *  @param[in] username     - the owner username of the session.
```
"the username of the session owner"

AlexanderAmelkin commented on include/libobmcsession/obmcsession.h:
```diff
@@ -0,0 +1,181 @@
+ *  @param[in] bus          - handle to system dbus
+ *  @param[in] slug         - the slug of service that is managing target
+ *                            session.
+ *  @param[in] username     - the owner username of the session.
```
same here

AlexanderAmelkin commented on include/libobmcsession/obmcsession.h:
```diff
@@ -0,0 +1,181 @@
+extern unsigned int obmcsesRemoveAll();
+
+/**
+ * @brief Get transaction status of session building
```
"session creation" ?

AlexanderAmelkin commented on include/libobmcsession/obmcsession.h:
```diff
@@ -0,0 +1,181 @@
+ * @brief Reset the active session build transaction
+ *
+ */
+extern void obmcsesResetPendginSessilBuild();
```
Pendgin? Sessil? Who are those people? ;)

AlexanderAmelkin commented on include/libobmcsession/obmcsession.hpp:
```diff
@@ -0,0 +1,77 @@
+ * @brief Create a session and publish into the dbus without appropriate
+ *        session payload
+ *
+ * @param[in] cleanupFn             - the callback to cleanup session on
```
comments made to .h file also apply here

AlexanderAmelkin commented on include/libobmcsession/obmcsession_proto.h:
```diff
@@ -0,0 +1,32 @@
+#define OBMC_SES_TYPE_REDFISH         (4U)
+#define OBMC_SES_TYPE_VIRTUAL_MEDIA   (5U)
+#define OBMC_SES_TYPE_WEB_UI          (6U)
+#define OBMC_SES_TYPE_NBD             (7U)
```
looks like a good candidate for an `enum class`.

ikmsk10 committed:
SSH: Add helper-service to control SSH sessions

End-user-impact: Now, the OpenBMC system has a way to provide
                 SSH-sessions information.

Signed-off-by: Igor Kononenko <i.kononenko@yadro.com>

AlexanderAmelkin commented:
Still need to address most of the previous comments
